phan 2.7 in core

phan 2.7 supports PHP 7.0 to 7.4 properly. 

- update phan to 2.7 in core
- adjust code and/or settings to make it pass

When switching from `phan` 1.3 to 2.7 we get the following results:
```
php -d zend.enable_gc=0 vendor-bin/phan/vendor/bin/phan --config-file .phan/config.php --require-config-exists -p
[info] Disabling Xdebug: Phan is around five times as slow when Xdebug is enabled (Xdebug only makes sense when debugging Phan itself)
[info] To run Phan with Xdebug, set the environment variable PHAN_ALLOW_XDEBUG to 1.
[info] To disable this warning, set the environment variable PHAN_DISABLE_XDEBUG_WARN to 1.
[info] To include function signatures of Xdebug, see .phan/internal_stubs/xdebug.phan_php
[debug] Checking PHAN_ALLOW_XDEBUG
[debug] Because xdebug was installed, Phan will restart.
[debug] The Xdebug extension is loaded (2.9.3)
[debug] Process restarting (PHAN_ALLOW_XDEBUG=internal|2.9.3|1|*|*)
[debug] Running '/usr/bin/php7.2' '-n' '-c' '/tmp/ZLtEXc' 'vendor-bin/phan/vendor/bin/phan' '--config-file' '.phan/config.php' '--require-config-exists' '-p'
   analyze ████████████████████████████████████████████████████████████ 100.0% 626MB/626MB
apps/federation/appinfo/routes.php:26 PhanUndeclaredThis Variable $this is undeclared
apps/files/appinfo/routes.php:32 PhanUndeclaredThis Variable $this is undeclared
apps/files/appinfo/routes.php:69 PhanUndeclaredThis Variable $this is undeclared
apps/files/appinfo/routes.php:71 PhanUndeclaredThis Variable $this is undeclared
apps/files/appinfo/routes.php:73 PhanUndeclaredThis Variable $this is undeclared
apps/files/appinfo/routes.php:76 PhanUndeclaredThis Variable $this is undeclared
apps/files_external/appinfo/routes.php:32 PhanUndeclaredThis Variable $this is undeclared
apps/files_external/appinfo/routes.php:50 PhanUndeclaredThis Variable $this is undeclared
apps/files_external/appinfo/routes.php:52 PhanUndeclaredThis Variable $this is undeclared
apps/files_external/appinfo/routes.php:55 PhanUndeclaredThis Variable $this is undeclared
apps/files_sharing/appinfo/routes.php:32 PhanUndeclaredThis Variable $this is undeclared
apps/files_sharing/appinfo/routes.php:143 PhanUndeclaredThis Variable $this is undeclared
apps/files_sharing/appinfo/routes.php:148 PhanUndeclaredThis Variable $this is undeclared
apps/files_sharing/appinfo/routes.php:150 PhanUndeclaredThis Variable $this is undeclared
apps/files_sharing/appinfo/routes.php:152 PhanUndeclaredThis Variable $this is undeclared
apps/files_sharing/appinfo/routes.php:154 PhanUndeclaredThis Variable $this is undeclared
apps/files_trashbin/appinfo/routes.php:28 PhanUndeclaredThis Variable $this is undeclared
apps/files_trashbin/appinfo/routes.php:30 PhanUndeclaredThis Variable $this is undeclared
apps/files_trashbin/appinfo/routes.php:32 PhanUndeclaredThis Variable $this is undeclared
apps/files_trashbin/appinfo/routes.php:34 PhanUndeclaredThis Variable $this is undeclared
apps/files_trashbin/appinfo/routes.php:36 PhanUndeclaredThis Variable $this is undeclared
apps/updatenotification/appinfo/routes.php:26 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:35 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:69 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:72 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:75 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:78 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:81 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:85 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:97 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:107 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:111 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:115 PhanUndeclaredThis Variable $this is undeclared
core/routes.php:121 PhanUndeclaredThis Variable $this is undeclared
lib/private/Comments/Manager.php:116 PhanTypeMismatchArgumentReal Argument 1 ($dateTime) is null of type null but \OCP\Comments\IComment::setLatestChildDateTime() takes \DateTime defined at lib/public/Comments/IComment.php:210
settings/Controller/UsersController.php:435 PhanAccessPropertyPrivate Cannot access private property \OC\Group\Manager->userManager defined at lib/private/Group/Manager.php:70
settings/Controller/UsersController.php:447 PhanAccessPropertyPrivate Cannot access private property \OC\Group\Manager->eventDispatcher defined at lib/private/Group/Manager.php:91
settings/Controller/UsersController.php:454 PhanAccessPropertyPrivate Cannot access private property \OC\Group\Manager->userManager defined at lib/private/Group/Manager.php:70
settings/routes.php:37 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:79 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:84 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:86 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:88 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:90 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:93 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:96 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:99 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:102 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:104 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:106 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:108 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:110 PhanUndeclaredThis Variable $this is undeclared
settings/routes.php:113 PhanUndeclaredThis Variable $this is undeclared
[debug] Restarted process exited 1
Makefile:227: recipe for target 'test-php-phan' failed
make: *** [test-php-phan] Error 1
```

--- a/.phan/config.php
+++ b/.phan/config.php
@@ -31,6 +31,17 @@ return [
 		'settings'
 	],
 
+	'exclude_file_list' => [
+		'apps/federation/appinfo/routes.php',
+		'apps/files/appinfo/routes.php',
+		'apps/files_external/appinfo/routes.php',
+		'apps/files_sharing/appinfo/routes.php',
+		'apps/files_trashbin/appinfo/routes.php',
+		'apps/updatenotification/appinfo/routes.php',
+		'core/routes.php',
+		'settings/routes.php'
+	],
+
 	// A directory list that defines files that will be excluded
 	// from static analysis, but whose class and method
 	// information should be included.

--- a/lib/private/Comments/Comment.php
+++ b/lib/private/Comments/Comment.php
@@ -302,7 +302,7 @@ class Comment implements IComment {
 	/**
 	 * sets the date of the most recent child
 	 *
-	 * @param \DateTime $dateTime
+	 * @param \DateTime|null $dateTime
 	 * @return IComment
 	 * @since 9.0.0
 	 */

--- a/lib/private/Comments/Manager.php
+++ b/lib/private/Comments/Manager.php
@@ -113,7 +113,7 @@ class Manager implements ICommentsManager {
 		if ($comment->getId() === '') {
 			$comment->setChildrenCount(0);
 			$comment->setLatestChildDateTime(new \DateTime('0000-00-00 00:00:00', new \DateTimeZone('UTC')));
-			$comment->setLatestChildDateTime(null);
+			$comment->setLatestChildDateTime();
 		}
 
 		if ($comment->getCreationDateTime() === null) {

--- a/lib/public/Comments/IComment.php
+++ b/lib/public/Comments/IComment.php
@@ -203,11 +203,11 @@ interface IComment {
 	/**
 	 * sets the date of the most recent child
 	 *
-	 * @param \DateTime $dateTime
+	 * @param \DateTime|null $dateTime
 	 * @return IComment
 	 * @since 9.0.0
 	 */
-	public function setLatestChildDateTime(\DateTime $dateTime);
+	public function setLatestChildDateTime(\DateTime $dateTime = null);
 
 	/**
 	 * returns the object type the comment is attached to

--- a/settings/Controller/UsersController.php
+++ b/settings/Controller/UsersController.php
@@ -432,6 +432,7 @@ class UsersController extends Controller {
 			}
 		}
 
+		'@phan-var \OC\User\Manager $this->userManager';
 		if ($this->userManager->userExists($username)) {
 			return new DataResponse(
 				[
@@ -483,6 +484,7 @@ class UsersController extends Controller {
 				$user->setEMailAddress($email);
 
 				try {
+					'@phan-var \OC\Settings\Controller\UsersController $this';
 					$this->generateTokenAndSendMail($username, $email);
 				} catch (\Exception $e) {
 					$this->log->error("Can't send new user mail to $email: " . $e->getMessage(), ['app' => 'settings']);

--- a/vendor-bin/phan/composer.json
+++ b/vendor-bin/phan/composer.json
@@ -1,5 +1,5 @@
 {
     "require": {
-        "phan/phan": "^1.3"
+        "phan/phan": "^2.7"
     }
 }

